Fix error for IPW Kaplan Meier

pzivich opened the pull request:
Finally got around to checking KM for propensity scores/inverse probability weights. When running the KM fitter, if weights=df['weights'] is not converted to a np.array, then unexpected behaviors occur. I have rewritten survival_table_from_events to convert the specified weights to a np.asarray(). Without this change the user would have to specify weights=np.asarray(df['weights']). Results were compared to SAS and consistent results were obtained with this fix

CamDavidsonPilon commented:
> then unexpected behaviors occur

can you detail what these may be? I'm curious what was missed in the tests: https://github.com/CamDavidsonPilon/lifelines/blob/master/tests/test_estimation.py#L495-L515

pzivich commented:
It relates to the index of the input dataframe/series. For example,
`
df = pd.DataFrame()
df['y'] = [0,0,1,1]
df['t'] = [0.816,0.805,0.571,0.990]
df['w'] = [1.001,1.005,1.000,0.992]
km = ll.KaplanMeierFitter()
km.fit(durations=df['t'],event_observed=df['y'],weights=df['w'])
print(km.survival_function_)
print(km.event_table)
`
Produces 
`          KM_estimate
timeline             
0.000        1.000000
0.571        0.749875
0.805        0.749875
0.816        0.749875
0.990        0.000000
          removed  observed  censored  entrance  at_risk
event_at                                                
0.000       0.000     0.000       0.0     3.998    3.998
0.571       1.000     1.000       0.0     0.000    3.998
0.805       1.005     0.000       1.0     0.000    2.998
0.816       1.001     0.000       1.0     0.000    1.993
0.990       0.992     0.992       0.0     0.000    0.992
`
as would be expected. However, if we make a slight change to the index of the input dataframe
`df = pd.DataFrame(index=[5,6,7,8])
df['y'] = [0,0,1,1]
df['t'] = [0.816,0.805,0.571,0.990]
df['w'] = [1.001,1.005,1.000,0.992]
km = ll.KaplanMeierFitter()
km.fit(durations=df['t'],event_observed=df['y'],weights=df['w'])
print(km.survival_function_)
print(km.event_table)
`
we instead get
`          KM_estimate
timeline             
0.000             1.0
0.571             1.0
0.805             1.0
0.816             1.0
0.990             1.0
          removed  observed  censored  entrance  at_risk
event_at                                                
0.000         0.0       0.0       0.0       0.0      0.0
0.571         0.0       0.0       0.0       0.0      0.0
0.805         0.0       0.0       0.0       0.0      0.0
0.816         0.0       0.0       0.0       0.0      0.0
0.990         0.0       0.0       0.0       0.0      0.0
`
which is not true. This can be fixed by instead inputting the weights series as a numpy array as such
`km = ll.KaplanMeierFitter()
km.fit(durations=df['t'],event_observed=df['y'],weights=np.asarray(df['w']))
print(km.survival_function_)
print(km.event_table)
`
producing the correct values
`          KM_estimate
timeline             
0.000        1.000000
0.571        0.749875
0.805        0.749875
0.816        0.749875
0.990        0.000000
          removed  observed  censored  entrance  at_risk
event_at                                                
0.000       0.000     0.000       0.0     3.998    3.998
0.571       1.000     1.000       0.0     0.000    3.998
0.805       1.005     0.000       1.0     0.000    2.998
0.816       1.001     0.000       1.0     0.000    1.993
0.990       0.992     0.992       0.0     0.000    0.992
`
I think this occurs due to how pandas is handling the series. I think it is trying to add those items to the newly created data in line 254-255 by the index. The tests did not catch this since they created dataframes with indexes that wouldn't break the code in this way. They started the index is default order, unlike above

CamDavidsonPilon commented:
Right, I've seen this "index" problem a few times. PR looks good - do you have time to add a test as well? (basically the code you have written above, making sure a non-trivial index works as expected)

pzivich commented:
Sure thing. I will add later today

CamDavidsonPilon commented on tests/test_estimation.py:
```diff
@@ -514,6 +514,19 @@ def test_weights_can_be_floats(self):
             assert True
 
+    def test_weights_with_unaligned_index(self):
+        pd.DataFrame(index=[5,6,7,8])
```
rm?

CamDavidsonPilon commented:
One comment, but overall lgtm! Thanks!

CamDavidsonPilon commented on tests/test_estimation.py:
```diff
@@ -514,6 +514,19 @@ def test_weights_can_be_floats(self):
+        df['w'] = [1.5,2,0.8,0.9]
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            k.fit(durations=df['t'],event_observed=df['y'],weights=df['w'])
```
`k` is not defined

pzivich commented:
My bad. Extra dataframe bit is deleted and fixed the call to the KaplanMeierFitter
